HDDS-4320. Let Ozone input streams implement CanUnbuffer

adoroszlai opened the pull request:
## What changes were proposed in this pull request?

Implement the [`CanUnbuffer` interface](https://github.com/apache/hadoop/blob/b32926f1108bef1f1e506de684c021203b2432f1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/CanUnbuffer.java#L32-L35) in Ozone input streams (`ChunkInputStream`, `BlockInputStream`): release buffers and disconnect the client when `unbuffer()` is called.  Upper level input streams (`KeyInputStream`, `OzoneInputStream`, `OzoneFSInputStream`) just delegate to the the underlying implementations.  This will allow Impala to [cache file handles](https://issues.apache.org/jira/browse/IMPALA-10214) for Ozone.

https://issues.apache.org/jira/browse/HDDS-4320

## How was this patch tested?

Added contract test (base test copied from Hadoop 3.3) for both OFS and O3FS:

https://github.com/adoroszlai/hadoop-ozone/runs/1309754540#step:4:3093
https://github.com/adoroszlai/hadoop-ozone/runs/1309754540#step:4:3107

also added unit test:

https://github.com/adoroszlai/hadoop-ozone/runs/1309754295#step:3:3540

adoroszlai commented:
Thanks @bshashikant for taking a look.

> 1. In unbuffer, do we need to remove the corresponding blockInputStreams and chunkInputStreams which are already read?

I don't think they should be removed.  They do not seem to add significant overhead.  The stream can be `seek`-ed both forward and backward.

> 2. While the connection to datanode is closed, connection to OM still is kept open, do we need to close this as well?

As far as I see, connection to OM is in the FileSystem instance (or rather the `OzoneClientAdapter` implementation), which is longer lived than the InputStreams (and does not support unbuffer).

> 3. As i see the last position is cached, so after unbuffer is called, it will go into corresponding blockInputStream and chunkInputStream and starts reading again. What if the pipelne is not valid anymore i.e, the datanodes containing the blocks is replicated to a different set of datanodes? Do we need to handle this?

Ideally we should.  But is this really specific to unbuffering?  I would think it can happen during a regular read, too.

bshashikant commented:
> Thanks @bshashikant for taking a look.
> 
> > 1. In unbuffer, do we need to remove the corresponding blockInputStreams and chunkInputStreams which are already read?
> 
> I don't think they should be removed. They do not seem to add significant overhead. The stream can be `seek`-ed both forward and backward.
> 
> > 1. While the connection to datanode is closed, connection to OM still is kept open, do we need to close this as well?
> 
> As far as I see, connection to OM is in the FileSystem instance (or rather the `OzoneClientAdapter` implementation), which is longer lived than the InputStreams (and does not support unbuffer).
> 
> > 1. As i see the last position is cached, so after unbuffer is called, it will go into corresponding blockInputStream and chunkInputStream and starts reading again. What if the pipelne is not valid anymore i.e, the datanodes containing the blocks is replicated to a different set of datanodes? Do we need to handle this?
> 
> Ideally we should. But is this really specific to unbuffering? I would think it can happen during a regular read, too.

Thanks @adoroszlai . I think with unbufffering , we would need to refresh the pipeline. Usually, while truing to read a block. we  first try to read the chunkInfo from datanodes during BlocknInputStream initialize() , and if that fails, we try to refresh the pipeline in case of regular read.

With unbuffer, connections to datanodes is closed, and the data will be the probably be read at later point of time. In cases as such, we might need to reinitialize BlockInputStream again.

adoroszlai commented:
> Usually, while trying to read a block. we first try to read the chunkInfo from datanodes during BlocknInputStream initialize() , and if that fails, we try to refresh the pipeline in case of regular read.

That's right.  But after initialization there is no guarantee that the pipeline will be there until all data is read.  Failures may happen any time.

> With unbuffer, connections to datanodes is closed, and the data will be the probably be read at later point of time. In cases as such, we might need to reinitialize BlockInputStream again.

I think always reinitializing the stream on first read after unbuffer would eliminate performance advantages of unbuffering, the client might as well close and open the file again.  Otherwise we need to come up with a list of specific errors, result codes that trigger reinitialization if encountered during read (while all other errors should still be propagated to the caller).  Do you have a suggestion for this of errors/result codes?

bshashikant commented:
> > Usually, while trying to read a block. we first try to read the chunkInfo from datanodes during BlocknInputStream initialize() , and if that fails, we try to refresh the pipeline in case of regular read.
> 
> That's right. But after initialization there is no guarantee that the pipeline will be there until all data is read. Failures may happen any time.
> 
> > With unbuffer, connections to datanodes is closed, and the data will be the probably be read at later point of time. In cases as such, we might need to reinitialize BlockInputStream again.
> 
> I think always reinitializing the stream on first read after unbuffer would eliminate performance advantages of unbuffering, the client might as well close and open the file again. Otherwise we need to come up with a list of specific errors, result codes that trigger reinitialization if encountered during read (while all other errors should still be propagated to the caller). Do you have a suggestion for this of errors/result codes?

I think we can atleast try to refresh for error like CONTAINER_NOT_FOUND , where it means either the container replica is deleted, or disks have gone bad.

adoroszlai commented:
> I think we can atleast try to refresh for error like CONTAINER_NOT_FOUND , where it means either the container replica is deleted, or disks have gone bad.

Thanks for the input @bshashikant.  I've added an integration test which shuts down all original datanodes one after another, and confirms that it can continue reading from a new datanode (after the container is replicated to that).  This shutdown results in an ugly chain of `IOException: ExecutionException: StatusRuntimeException: UNAVAILABLE: io exception`.  To handle it, I've updated the patch to refresh on all IO exceptions, not just container not found.  It will propagate the exception if the refresh does not result in a new pipeline.

hanishakoneru commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/contract/AbstractContractUnbufferTest.java:
```diff
@@ -0,0 +1,159 @@
+
+  @Test
+  public void testUnbufferOnClosedFile() throws IOException {
+    describe("unbuffer a file before a read");
```
typo in description

hanishakoneru commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/contract/AbstractContractUnbufferTest.java:
```diff
@@ -0,0 +1,159 @@
+    long pos = stream.getPos();
+    stream.unbuffer();
+    assertEquals("unbuffer unexpectedly changed the stream position", pos,
+            stream.getPos());
```
unbuffer() here just checks that the position is maintained. Would it be possible to also verify that the buffers are actually released?

adoroszlai replied:
I don't think it can, as it does not know about implementation internals.  This is verified in `TestChunkInputStream` instead.  Hadoop also has additional implementation-specific test for this reason.

hanishakoneru reviewed (commented):
Thanks @adoroszlai for working on this. The patch LGTM. 
I have some minor comments. I am good with the logic and implementation part.

adoroszlai commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:
```diff
@@ -171,6 +162,23 @@ public synchronized void initialize() throws IOException {
 
+  private void refreshPipeline(IOException cause) throws IOException {
+    LOG.error("Unable to read information for block {} from pipeline {}: {}",
+        blockID, pipeline.getId(), cause.getMessage());
```
It was an error previously, I only changed the message a bit from:

```
LOG.error("Unable to read block information from pipeline.");
```

but I agree, it should be logged at lower level.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestKeyInputStream.java:
```diff
@@ -397,4 +410,118 @@ public void testSkip() throws Exception {
+      cluster.shutdownHddsDatanode(pipelineNodes.get(0));
+      waitForNodeToBecomeDead(pipelineNodes.get(0));
+      waitForReplicaCount(containerID, 2, cluster);
+      waitForReplicaCount(containerID, 3, cluster);
```
We need to waitForReplicaCount=2 first to ensure one of the original replicas is lost.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/contract/AbstractContractUnbufferTest.java:
```diff
@@ -0,0 +1,159 @@
+    byte[] streamData = new byte[length];
+    assertEquals("failed to read expected number of bytes from "
+            + "stream. This may be transient",
+        length, stream.read(streamData));
```
`AbstractContractUnbufferTest` is copied as is from Hadoop 3.3 source, since we cannot yet upgrade dependency to that version.  The only changes I had to make were for checkstyle (too long line) and the new constant `SUPPORTS_UNBUFFER` introduced in `ContractOptions` (which the test implements).

adoroszlai commented:
Thanks @hanishakoneru for the review.  I have updated some names and docs according to your comments.

hanishakoneru commented:
Thanks @adoroszlai. LGTM. +1.

adoroszlai commented:
Thanks @bshashikant and @hanishakoneru for the reviews.
